ksql-engine/src/main/java/io/confluent/ksql/planner/plan/KsqlStructuredDataOutputNode.java

```diff
@@ -171,4 +224,39 @@ public KeyField getKeyField() {
 
     return result.selectKey(newKey.name(), false, contextStacker);
   }
+
+  private static Set<Integer> implicitAndKeyColumnIndexes(final LogicalSchema schema) {
```

Comment:
reading this was very confusing to me - I had no idea why we would want a set of columns indices that include both ROWTIME and any key fields... it seems like it's for "serialization" purposes (making sure the keys aren't duplicated into the value). IMO, this logic belongs somewhere else and we should be passing down the LogicalSchema to the place that needs this information (looks like `SchemaKStream`). That logic would then look like:
```java
for (int i = 0; i < row.getColumns().size(); i++) {
   if (!logicalSchema.isKey(i) && !logicalSchemas.isImplicit(i)) {
      columns.add(...);
   }
}
```

Reply:
Humm... so... this logical was originally in `SchemaUtil`.  I moved it to `LogicalSchema` when I created the class. However, it's only used in one place: by `KsqlStructuredDataOutputNode`. Hence I've moved the logic to that class. 

>  I had no idea why we would want a set of columns indices that include both ROWTIME and any key fields

_Why_ this is needed has not changed with the change I'm making. It wasn't particularly self explanatory before, and, agreed, it still isn't. 

This whole area needs improving as the current method has many strange edge cases and artefacts. However, that's not something I'm doing in this PR. What this PR is trying to do is move this specific bit of existing logic from `LogicalSchema` to the only place its used, thereby cleaning up the `LogicalSchema` API.  With a view to improving how the whole thing works later.

As per the to-do in the code, the final code here will make use of the `implicitFields` and `keyFields` methods of the `LogicalSchema` class once the other PR is merged.  

Your  alternative suggests you haven't yet got your head around what the code is doing. Which is in itself not a great sign! ;) . But as I've said, I'm not trying to change the functionality in this PR, only move it.

Your alternative, with `boolean isKey(int columnIndex)` style methods suggests you think this method is working out which of the fields in the logical schema are ROWTIME or ROWKEY.   That is not the case.  The method is finding the indexes of ROWTIME and ROWKEY fields _in the value schema only_.  (I've renamed this method to `implicitAndKeyColumnIndexesInValueSchema` in the hope this will help make this clear).  This is existing functionality. It's needed so that KSQL can remove the ROWKEY and ROWTIME columns its copied into the value schema in `DataSourceNode` are removed and not persisted in the value of the output.  This will need to change before we can do structured keys and if we want to support quoted identifiers e.g. 

```sql
CREATE STREAM AS SELECT `RowKey`, x AS `ROWTIME` from Foo;
```

Depending on how quoted identifiers is implemented, the above won't work as `KsqlStructuredDataOutputNode` will filter out these two columns. 

But this is all something for another PR.

Hope this helps explain why it is like it is. But the bottom line is, all I've done is _move_ a function.

Reply:
> It's needed so that KSQL can remove the ROWKEY and ROWTIME columns its copied into the value schema in DataSourceNode are removed and not persisted in the value of the output. 

I believe this is the way I understood it, but maybe I'm still not getting anything. Either way, thanks for the detailed explanation - this PR is definitely no worse than what it was so if it helps you in your master plan I'm happy to oblige :)